### PR TITLE
Bump UMF version and switch to dynamic linking

### PR DESCRIPTION
**Pull request by igchor**

### CODE REVIEW

**Comment by igchor** [comment]
> Let's target these changes for after 0.9.

Sure, I just wanted to create a PR to do some initial testing

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -35,7 +35,7 @@ jobs:
       - name: Install apt package
         run: |
           sudo apt-get update
-          sudo apt-get install -y doxygen
+          sudo apt-get install -y libhwloc-dev doxygen
```
Comment (PatKamin) [inline .github/workflows/docs.yml:38]: hwloc not needed for generating docs.

File: .github/workflows/cmake.yml [hunk 2]
```diff
@@ -240,6 +245,11 @@ jobs:
 
     - name: Install pip packages
       run: pip install -r third_party/requirements.txt
+
+    - name: Install apt packages
```
Comment (lukaszstolarczuk) [inline .github/workflows/cmake.yml:249]: this job is run on our HW - it should not be needed here

Comment (igchor) [reply .github/workflows/cmake.yml]: Done.

File: .github/workflows/docs.yml (hunk 1 above)
Comment (igchor) [reply .github/workflows/docs.yml]: Done.

**Review by lukaszstolarczuk — COMMENTED** [review]
👍

**Comment by igchor** [comment]
OpenCL e2e test failure are unrelated to the PR (they fail on main as well).

**Comment by igchor** [comment]
There is one temporary commit on this PR to run e2e tests on my fork of intel/llvm. This is because I've changed how the common library is linked to the adapters/plugins and this requires changes in llvm (as you can see here: https://github.com/intel/llvm/pull/13343)

**Comment by igchor** [comment]
I've removed the tmp commit which mean e2e tests will be failing until we merge changes to intel/llvm. Last successful build is here: https://github.com/oneapi-src/unified-runtime/actions/runs/9842229066/job/27204428422?pr=1430

**Comment by igchor** [comment]
@kbenzie once https://github.com/intel/llvm/pull/13343 passes, this should be ready to merge

File: .github/workflows/build-fuzz-reusable.yml [hunk 3]
```diff
@@ -33,6 +33,11 @@ jobs:
     - name: Install pip packages
       run: pip install -r third_party/requirements.txt
 
+    - name: Install apt packages
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y libhwloc-dev
```
Comment (lukaszstolarczuk) [inline .github/workflows/build-fuzz-reusable.yml:39]: this job runs now on self-hosted runner with no sudo - libhwloc should be there already - you can remove this

**Comment by kbenzie** [comment]
I guess we should wait to merge this until we have confirmation [here](https://github.com/intel/llvm/pull/13343#issuecomment-2223558946).

File: .github/scripts/install_hwloc.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+# Copyright (C) 2024 Intel Corporation
+# Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
```
Comment (lukaszstolarczuk) [inline .github/scripts/install_hwloc.sh:3]: UR license...?

File: .github/workflows/build-fuzz-reusable.yml (hunk 3 above)
Comment (kswiecicki) [reply .github/workflows/build-fuzz-reusable.yml]: Done.

File: .github/scripts/install_hwloc.sh (hunk 4 above)
Comment (kswiecicki) [reply .github/scripts/install_hwloc.sh]: Done.

**Review by PatKamin — APPROVED** [review]

**Comment by igchor** [comment]
@kbenzie we have confirmation that intel/llvm is ready for the change: https://github.com/intel/llvm/pull/13343#pullrequestreview-2198112559
@kswiecicki please bump the UR tag once https://github.com/oneapi-src/unified-memory-framework/pull/633 is merged and add `ready to merge` label

**Review by lukaszstolarczuk — COMMENTED** [review]
👍

**Review by pbalcer — APPROVED** [review]

**Comment by ldrumm** [comment]
This patch does more than two things and has no description.

Why are we switching to dynamic linking?
To which version are we bumping?
Which parts of the code changes are really required by the bump, and which are in-passing code cleanups?

Right now this patch does *everything* without rationale. Please split it if you can, remove the fly-by changes, and provide us with a description of what justifies such changes

**Review by kbenzie — APPROVED** [review]

**Pull request merged**